From 235949a844c35c7c1ec5ab2e14c1c28c2c47a083 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 26 Oct 2025 10:19:17 +0100
Subject: [PATCH 0993/1013] evl/net: tap: fix inconsistent lock usage when
 feeding nit

This fixes a lockdep splat observed when capturing packets with
tcpdump from an interface with an active oob port:

[  430.983437] inconsistent {IN-SOFTIRQ-W} -> {SOFTIRQ-ON-W} usage.
[  430.983443] kworker/3:1/73 [HC0[0]:SC0[0]:HE1:SE1] takes:
[  430.983457] c3f868c4 (rlock-AF_PACKET){+.?.}-{2:2}, at: tpacket_rcv+0x218/0xc28
[  430.983497] {IN-SOFTIRQ-W} state was registered at:
[  430.983503]   lock_acquire+0x130/0x378
[  430.983526]   _raw_spin_lock+0x40/0x50
[  430.983549]   tpacket_rcv+0x218/0xc28
[  430.983560]   __netif_receive_skb_core.constprop.0+0x5a0/0xd18
[  430.983589]   __netif_receive_skb_list_core+0x110/0x24c
[  430.983606]   netif_receive_skb_list_internal+0x21c/0x404
[  430.983617]   napi_complete_done+0xd8/0x1d8
[  430.983626]   fec_enet_rx_napi+0xac0/0x121c
[  430.983649]   __napi_poll.constprop.0+0x2c/0x26c
[  430.983659]   net_rx_action+0x1e8/0x3c4
[  430.983668]   handle_softirqs+0x124/0x4e8
[  430.983686]   __irq_exit_rcu+0x15c/0x1d4
[  430.983699]   irq_exit+0x8/0x28
[  430.983712]   arch_do_IRQ_pipelined+0x34/0x60
[  430.983725]   sync_current_irq_stage+0x134/0x310
[  430.983735]   __inband_irq_enable+0x4c/0x58
[  430.983744]   finish_task_switch+0xd4/0x32c
[  430.983768]   __schedule+0x468/0xf70
[  430.983780]   schedule_idle+0x20/0x64
[  430.983791]   cpu_startup_entry+0x28/0x2c
[  430.983814]   kernel_init+0x0/0x12c
[  430.983835] irq event stamp: 4205
[  430.983840] hardirqs last  enabled at (4205): [<c13d6f28>] _raw_spin_unlock_irq+0x20/0x38
[  430.983862] hardirqs last disabled at (4204): [<c13d6ca4>] _raw_spin_lock_irq+0x24/0x70
[  430.983880] softirqs last  enabled at (4198): [<c04ca954>] skb_recycler+0xe4/0x120
[  430.983900] softirqs last disabled at (4088): [<c04ca8e0>] skb_recycler+0x70/0x120
[  430.983915]
[  430.983915] other info that might help us debug this:
[  430.983919]  Possible unsafe locking scenario:
[  430.983919]
[  430.983923]        CPU0
[  430.983926]        ----
[  430.983929]   lock(rlock-AF_PACKET);
[  430.983937]   <Interrupt>
[  430.983940]     lock(rlock-AF_PACKET);
[  430.983948]
[  430.983948]  *** DEADLOCK ***
[  430.983948]
[  430.983952] 3 locks held by kworker/3:1/73:
[  430.983958]  #0: c30120b4 ((wq_completion)events){+.+.}-{0:0}, at: process_one_work+0x1b4/0x6b4
[  430.983990]  #1: f0cd5f10 ((work_completion)(&r->work)){+.+.}-{0:0}, at: process_one_work+0x1e0/0x6b4
[  430.984018]  #2: c2092324 (rcu_read_lock){....}-{1:2}, at: dev_queue_recv_nit+0x0/0x318
[  430.984057]
[  430.984057] stack backtrace:
[  430.984063] CPU: 3 UID: 0 PID: 73 Comm: kworker/3:1 Not tainted 6.17.0-00887-gc70cb95c32fd #241 NONE
[  430.984082] Hardware name: Freescale i.MX6 Quad/DualLite (Device Tree)
[  430.984088] IRQ stage: Linux
[  430.984097] Workqueue: events evl_net_tap_in__batch
[  430.984119] Call trace:
[  430.984125]  unwind_backtrace from show_stack+0x10/0x14
[  430.984154]  show_stack from dump_stack_lvl+0xa0/0xe0
[  430.984180]  dump_stack_lvl from print_usage_bug.part.0+0x2b8/0x358
[  430.984209]  print_usage_bug.part.0 from mark_lock.part.0+0xc28/0x12bc
[  430.984233]  mark_lock.part.0 from __lock_acquire+0xa4c/0x29f8
[  430.984256]  __lock_acquire from lock_acquire+0x130/0x378
[  430.984279]  lock_acquire from _raw_spin_lock+0x40/0x50
root@phytec-mira-evl:~# [  430.984307]  _raw_spin_lock from tpacket_rcv+0x218/0xc28
[  430.984330]  tpacket_rcv from dev_queue_recv_nit+0x180/0x318
[  430.984357]  dev_queue_recv_nit from evl_net_tap_in__batch+0x34/0x60
[  430.984387]  evl_net_tap_in__batch from process_one_work+0x258/0x6b4
[  430.984409]  process_one_work from worker_thread+0x1a4/0x3a4
[  430.984425]  worker_thread from kthread+0x118/0x268
[  430.984453]  kthread from ret_from_fork+0x14/0x20
[  430.984476] Exception stack(0xf0cd5fb0 to 0xf0cd5ff8)
[  430.984488] 5fa0:                                     00000000 00000000 00000000 00000000
[  430.984499] 5fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  430.984509] 5fe0: 00000000 00000000 00000000 00000000 00000013 00000000

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/tap.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/evl/net/tap.c b/kernel/evl/net/tap.c
index 292df450dd59..61bbf23fbeff 100644
--- a/kernel/evl/net/tap.c
+++ b/kernel/evl/net/tap.c
@@ -28,7 +28,9 @@ static void feed_tap_in(struct net_device *dev, struct sk_buff *skb)
 
 static void deliver_input_nit(struct evl_net_tap_data *data)
 {
+	local_bh_disable();
 	feed_tap_in(data->dev, data->skb);
+	local_bh_enable();
 	evl_net_free_skb(data->skb);
 }
 
-- 
2.30.2

